File: lib/src/lidar/lidar.dart
import "dart:developer";
import "dart:ffi";

import "package:video/src/collection.dart";

import "../generated/lidar_ffi_bindings.dart";
import "../generated/lidar_bindings.dart";
import "package:ffi/ffi.dart";
import "package:opencv_ffi/opencv_ffi.dart";

const cliArgs = "hostname:=169.254.166.55";
const launchFile = "lidar.launch";

typedef NativePointCloudMsgCallback = Void Function(
    Pointer<Void>, Pointer<SickScanPointCloudMsgType>);

typedef SickScanDiagnosticMsgCallback = Void Function(
  Pointer<Void>, Pointer<SickScanDiagnosticMsg>);

/// An ffi implementation of SICK Lidar API
class LidarFFI {
  final LidarBindings bindings;
  late Pointer<Void> _handle;
  OpenCVImage? lastestImage = null;
  late final NativeCallable<NativePointCloudMsgCallback> callback;
  late final NativeCallable<SickScanDiagnosticMsgCallback> errorCallback;
  LidarFFI() : bindings = LidarBindings(DynamicLibrary.open("lidar.dll"));

  Future<bool> init() => using((arena) async {
        _handle = bindings.SickScanApiCreate(0, nullptr);
        // bindings.SickScanApiInitByLaunchfile(_handle, filename);
        // final array = arena<Char>();
        final argsPtr = arena<Pointer<Char>>(3);
        argsPtr[0] = "lidar.dart".toNativeUtf8(allocator: arena).cast<Char>();
        argsPtr[1] = launchFile.toNativeUtf8(allocator: arena).cast<Char>();
        argsPtr[2] = cliArgs.toNativeUtf8(allocator: arena).cast<Char>();

        final result = bindings.SickScanApiInitByCli(_handle, 3, argsPtr);

        // This is the actual callback function
        // TODO: need to make this do something
        // TODO: put somewhere else in class
        // TODO: Get rid o

        void handler(SickScanApiHandle apiHandle, Pointer<SickScanPointCloudMsg> msg) {
          // Remember to close the NativeCallable once the native API is
          // finished with it, otherwise this isolate will stay alive
          // indefinitely.
          // struct ScikScanblajh p = ;'
          // print("Message address: ${msg}");
          if (msg.ref.height == 0 || msg.ref.width == 0) {
            logger.warning("If zero"); 
            return;
          }

          print("Message address: ${msg}");

          print("Recieved msg with size: ${msg.ref.height} x ${msg.ref.width}");
          print("Types ${msg.ref.height.runtimeType} x ${msg.ref.width.runtimeType}");
          print("is big endian? ${msg.ref.is_bigendian}");
          print("point step: ${msg.ref.point_step}");
          print("row step: ${msg.ref.row_step}");
          //print("Buffer address: ${msg.ref.fields.buffer}");

          return using(
            (arena) async {
              //final struct = arena<SickScanPointCloudMsgType>();
              //final fieldBuffer = msg.ref.fields.buffer;
              print("data capacity: ${msg.ref.data.capacity}, size: ${msg.ref.data.size}");
              print("fields capacity: ${msg.ref.fields.capacity}, size: ${msg.ref.fields.size}");
              //print("Buffer address: ${msg.ref.fields.buffer}");
              print("${msg.ref.fields.buffer.ref.count}");
              print("herllokfe");
              int field_offset_x = -1,
                  field_offset_y = -1,
                  field_offset_z = -1,
                  field_offset_intensity = -1;


              // for (int i = 0; i < msg.ref.fields.size; i++) {
              //   print("hello + ${fieldBuffer[i].name}");
              //   final name = fieldBuffer[i].name;
              //   print(name);
              //   final datatype =
              //       SickScanNativeDataType.fromValue(fieldBuffer[i].datatype);
              //   print(datatype);
              //   final value = fieldBuffer[i].offset;
              //   print("a 1 [$name] $datatype $value");
              //   print(name);
              //   print(datatype ==
              //       SickScanNativeDataType
              //           .SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32);

              //   if (name == "x" &&
              //       datatype ==
              //           SickScanNativeDataType
              //               .SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32) {
              //     field_offset_x = fieldBuffer[i].offset;
              //     // print("  here x");
              //   } else if (name == "y" &&
              //       datatype ==
              //           SickScanNativeDataType
              //               .SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32) {
              //     field_offset_y = fieldBuffer[i].offset;
              //     // print("  here y");
              //   } else if (name == "z" &&
              //       datatype ==
              //           SickScanNativeDataType
              //               .SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32) {
              //     field_offset_z = fieldBuffer[i].offset;
              //     // print("  here z");
              //   } else if (name == "intensity" &&
              //       datatype ==
              //           SickScanNativeDataType
              //               .SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32) {
              //     field_offset_intensity = fieldBuffer[i].offset;
              //     // print("  here i");
              //   }
              // }
              // if (field_offset_x < 0 ||
              //     field_offset_y < 0 ||
              //     field_offset_z < 0) {
              //   print("fields wrong");
              //   return null;
              // }

              // const imgWidth = 250 * 4;
              // const imgHeight = 250 * 4;

              // C Code uint8_t* pixels = (uint8_t*)calloc(3 * img_width * img_height, sizeof(uint8_t));
              // final pixels = arena<Uint8>(3 * imgWidth * imgHeight);
              // addHiddenArea(imgHeight, imgWidth, pixels);
              // for (var row_idx = 0; row_idx < msg.ref.height; row_idx++) {
              //   for (var col_idx = 0; col_idx < msg.ref.width; col_idx++) {
              //     // Get cartesian point coordinates
              //     var polar_point_offset =
              //         row_idx * msg.ref.row_step + col_idx * msg.ref.point_step;
              //     var address = msg.ref.data.buffer.address +
              //         polar_point_offset +
              //         field_offset_x;
              //     final point_x = Pointer<Float>.fromAddress(address).value;
              //     address = msg.ref.data.buffer.address +
              //         polar_point_offset +
              //         field_offset_y;
              //     final point_y = Pointer<Float>.fromAddress(address).value;
              //     address = msg.ref.data.buffer.address +
              //         polar_point_offset +
              //         field_offset_z;
              //     final point_z = Pointer<Float>.fromAddress(address).value;
              //     double point_intensity = 0;
              //     if (field_offset_intensity >= 0) {
              //       address = msg.ref.data.buffer.address +
              //           polar_point_offset +
              //           field_offset_intensity;
              //       point_intensity = Pointer<Float>.fromAddress(address).value;
              //     }
              //     // print("a2.5");

              //     // Convert point coordinates in meter to image coordinates in pixel
              //     int img_x = (250.0 * (-point_y + 2.0))
              //         .toInt(); // img_x := -pointcloud.y
              //     int img_y = (250.0 * (-point_x + 2.0))
              //         .toInt(); // img_y := -pointcloud.x
              //     if (img_x >= 0 &&
              //         img_x < imgWidth &&
              //         img_y >= 0 &&
              //         img_y < imgHeight) // point within the image area
              //     {
              //       pixels[3 * img_y * imgWidth + 3 * img_x + 0] = 0; // B
              //       pixels[3 * img_y * imgWidth + 3 * img_x + 1] = 255; // G
              //       pixels[3 * img_y * imgWidth + 3 * img_x + 2] = 255; // R
              //     }
              //   }
              // }

              // addCross(imgHeight, imgWidth, pixels);
              // print("GOT THIS FAR");
              // final mat = getMatrix(imgHeight, imgWidth, pixels);
              // final jpeg = encodeJpg(mat);
              // print("  IMAGE IS: $jpeg");
              // if (jpeg == null) {
              //   print("null image");
              //   return null;
              // }
              //lastestImage = jpeg;
              // bindings.SickScanApiFreePointCloudMsg(_handle, msg);
            },
          );
          //callback.close();
        }
        void DiagnosticMsg(SickScanApiHandle apiHandle, Pointer<SickScanDiagnosticMsg> msg){
          if(msg.ref.status_code == 1){
            print("${msg.ref.status_code} WARNING");
          }
          else if(msg.ref.status_code == 2){
            print("${msg.ref.status_code} ERROR");
          }
          else{
            print("${msg.ref.status_code} Other issue");
          }
          
          return using((arena) async{
            final Pointer<Char> pointerToMsgBuffer = arena<Char>(1024);
            final statusCode = arena<Int32>();
            statusCode.value = -1;
            if(bindings.SickScanApiGetStatus(apiHandle, statusCode, pointerToMsgBuffer, 1024) == 0){
              logger.info("SickScanApiGetStatus(apiHandle:$apiHandle): status_code = $statusCode, message = $pointerToMsgBuffer");
            }
            else{
              logger.error("SickScanApiGetStatus(apiHandle:$apiHandle) failed");
            }

          });
          
          
        }

        callback =
            NativeCallable<NativePointCloudMsgCallback>.listener(handler);
        
        errorCallback = NativeCallable<SickScanDiagnosticMsgCallback>.listener(DiagnosticMsg);


        final bindings2 = LidarBinding(DynamicLibrary.open("lidar_ffi.dll"));
        bindings.SickScanApiRegisterCartesianPointCloudMsg(
            _handle, bindings2.addresses.updateLatestImage.cast());

        bindings.SickScanApiRegisterDiagnosticMsg(_handle, errorCallback.nativeFunction);

        await Future<void>.delayed(const Duration(seconds: 10));
        if (result != 0) {
          print("Unable to initialize Sick Scan Lidar");
          return false;
        }
        print("finished init");
        return true;
      });

  void dispose() {
    errorCallback.close();
    callback.close();
    bindings.SickScanApiClose(_handle);
    bindings.SickScanApiRelease(_handle);
  }

  

  /// Converts a [SickScanPointCloudMsg] into a single [OpenCVImage]
  Future<OpenCVImage?> getOneImage({double timeout = 5}) async => lastestImage;

  // await Future<void>.delayed(Duration(seconds: timeout.toInt()));
  // if(result != 0){
  //   print("There was a problem calling SickScanApiWaitNextCartesianPointCloudMsg()");
  //   return null;
  // }
  //final length = struct.ref.data.size;
  //print(struct.ref.data.buffer.asTypedList(length));
  // print(struct.ref.fields.size);
  // });

  /// Add a red cross to the center of an image
  void addCross(int imgHeight, int imgWidth, Pointer<Uint8> pixels,
      {int thickness = 1}) {
    final midx = imgWidth ~/ 2;
    final midy = imgHeight ~/ 2;
    for (var x = midx - 7; x <= midx + 7; x++) {
      // draw horizontal
      for (var y = midy - thickness; y < midy + thickness; y++) {
        pixels[3 * y * imgWidth + 3 * x + 0] = 0; // B
        pixels[3 * y * imgWidth + 3 * x + 1] = 0; // G
        pixels[3 * y * imgWidth + 3 * x + 2] = 255; // R
      }
    }
    for (var y = midy - 7; y <= midy + 7; y++) {
      // draw vertical
      for (var x = midx - thickness; x < midx + thickness; x++) {
        pixels[3 * y * imgWidth + 3 * x + 0] = 0; // B
        pixels[3 * y * imgWidth + 3 * x + 1] = 0; // G
        pixels[3 * y * imgWidth + 3 * x + 2] = 255; // R
      }
    }
  }

  /// Draws a triangle in the area behind lidar that doesn't include data
  void addHiddenArea(int imgHeight, int imgWidth, Pointer<Uint8> pixels) {
    /// NEED IMAGE TO BE SQUARE FOR THIS TO WORK
    for (var y = imgHeight - 1; y > imgHeight ~/ 2; y--) {
      for (var x = imgWidth - y; x < y; x++) {
        pixels[3 * y * imgWidth + 3 * x + 0] = 130; // R
        pixels[3 * y * imgWidth + 3 * x + 1] = 130; // G
        pixels[3 * y * imgWidth + 3 * x + 2] = 130; // B
      }
    }
  }
}

/// Extension on Array<Char>
extension ToDartString on Array<Char> {
  /// Converts an [Array] of [Char] to a native dart [String]
  String toDartString({int length = 256}) {
    print("hello1");

    var i = 0;
    final buffer = StringBuffer();
    print(buffer);
    while (i < length) {
      print("Buffer: $buffer , Buffer Size :${buffer.length} iteration: ${this[i]}");

      if (String.fromCharCode(this[i]) == "\n"){
        break;
      }
      if(this[i] == 0){
          break;
      }
      buffer.writeCharCode(this[i]);
      i++;
    }

    return buffer.toString();
  }
}

void main() async {
  final sensor = LidarFFI();
  print("Start init");
  sensor.init();
  await Future<void>.delayed(const Duration(seconds: 10));
  print("Finished init");
  /*
  for(var i = 0; i < 5; i++){
    final image = await sensor.getOneImage();
    print(image);
    await Future<void>.delayed(const Duration(seconds: 5));
  }
  */

  sensor.dispose();

  await Future<void>.delayed(Duration(seconds: 10));
  sensor.dispose();
  print("Finished Program Execution");
  return;
}

File: src/lidar/lidar.cpp
// #include <chrono>
#include <iostream>
// #include <thread>

// #include "lidar.h"

// Image* ReturnLatestImage(void* msg){
//     ReturnLatestImage * latestImage;
//     latestImage->buffer = msg->data;
//     latestImage->size = msg->height;
//     latestImage->capacity  = msg->width;

//     return latestImage;
// }
// static void exitOnError(const char* msg, int32_t error_code)
// {
//     printf("## ERROR sick_scan_xd_api_test: %s, error code %d\n", msg, error_code);
//     exit(EXIT_FAILURE);
// }

// static void apiTestCartesianPointCloudMsgCallback(SickScanApiHandle apiHandle, const SickScanPointCloudMsg* msg)
// {
//     printf("[Info]: apiTestCartesianPointCloudMsgCallback(apiHandle:%p): %dx%d pointcloud callback...\n", apiHandle, msg->width, msg->height);
// }

// int main(int argc, char** argv, const std::string& sick_scan_args, bool polling){
//   int32_t ret = SICK_SCAN_API_SUCCESS;
//   SickScanApiHandle apiHandle = 0;

//   if ((apiHandle = SickScanApiCreate(argc, argv)) == 0)
//     exitOnError("SickScanApiCreate failed", -1);

//   // Initialize a lidar and starts message receiving and processing
// #if __ROS_VERSION == 1
//   if ((ret = SickScanApiInitByLaunchfile(apiHandle, sick_scan_args.c_str())) != SICK_SCAN_API_SUCCESS)
//     exitOnError("SickScanApiInitByLaunchfile failed", ret);
// #else
//   if ((ret = SickScanApiInitByCli(apiHandle, argc, argv)) != SICK_SCAN_API_SUCCESS)
//     exitOnError("SickScanApiInitByCli failed", ret);
// #endif

// }

#include "lidar.h"
#include <cassert>

Image image;

SickScanApiHandle handle;

int mutex;

FFI_PLUGIN_EXPORT void init() {
  image.data = (uint8_t*)malloc(image.width*image.height*3*sizeof(uint8_t));
  handle = SickScanApiCreate(0, nullptr);
  SickScanApiRegisterCartesianPointCloudMsg(handle, updateLatestImage);
  SickScanApiSetVerboseLevel(handle, 0); // 0 = DEBUG
  char* args[] = {"lidar.dart", "lidar.launch", "hostname:=169.254.166.55"};
  SickScanApiInitByCli(handle, 3, args);
  // ALLOCATE MEMORY FOR image.data 
  std::cout << "INITEDDDDD !!!" << std::endl;
}

FFI_PLUGIN_EXPORT void dispose() {
  SickScanApiDeregisterCartesianPointCloudMsg(handle, updateLatestImage);
  SickScanApiClose(handle);
  SickScanApiRelease(handle);
}

FFI_PLUGIN_EXPORT void updateLatestImage(SickScanApiHandle apiHandle, const SickScanPointCloudMsg* pointCloudMsg) {
  std::cout << "Image height: " << (int) pointCloudMsg->height << ", Width: " << (int) pointCloudMsg->width << std::endl;
  // return;
  if(mutex == 0) return;
  mutex = 0;
  // Change to if: assert(pointCloudMsg->height >= 0 && (int)pointCloudMsg->width >=0);
  if((int)pointCloudMsg->height == 0 && (int)pointCloudMsg->width ==0){
    image.height = pointCloudMsg->height;
    image.width = pointCloudMsg->width;
    return;
  }
  image.height = pointCloudMsg->height;
  image.width = pointCloudMsg->width;
  if (image.data == nullptr) {
    image.data = new uint8_t[image.height * image.width * 3];
  }
  // make_matrix(pointCloudMsg);
  // addCross(pointCloudMsg);
  addHiddenArea();
  mutex = 1;
}

FFI_PLUGIN_EXPORT void make_matrix(SickScanPointCloudMsg* imageData){
  SickScanPointFieldMsg* imageData_fields_buffer = (SickScanPointFieldMsg*)imageData->fields.buffer;
  int field_offset_x = -1, field_offset_y = -1;
  for(int n = 0; n < imageData->fields.size; n++)
    {
        if (strcmp(imageData_fields_buffer[n].name, "x") == 0 && imageData_fields_buffer[n].datatype == SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32)
            field_offset_x = imageData_fields_buffer[n].offset;
        else if (strcmp(imageData_fields_buffer[n].name, "y") == 0 && imageData_fields_buffer[n].datatype == SICK_SCAN_POINTFIELD_DATATYPE_FLOAT32)
            field_offset_y = imageData_fields_buffer[n].offset;
    }
  assert(field_offset_x >= 0 && field_offset_y >= 0);
  int img_width = 250 * 4, img_height = 250 * 4;

  for(int row = 0; row < (int)imageData->height; row++){
    for(int col = 0; col < (int)imageData->width; col++){
    int polar_point_offset = row * imageData->row_step + col * imageData->point_step;
    float point_x = *((float*)(imageData->data.buffer + polar_point_offset + field_offset_x));
    float point_y = *((float*)(imageData->data.buffer + polar_point_offset + field_offset_y));
			// Convert point coordinates in meter to image coordinates in pixel
			int img_x = (int)(250.0f * (-point_y + 2.0f)); // img_x := -pointcloud.y
			int img_y = (int)(250.0f * (-point_x + 2.0f)); // img_y := -pointcloud.x
			if (img_x >= 0 && img_x < img_width && img_y >= 0 && img_y < img_height) // point within the image area
			{
        std::cout << "Before" << std::endl;
				image.data[3 * img_y * img_width + 3 * img_x + 0] = 255; // R
        std::cout << "After" << std::endl;

				image.data[3 * img_y * img_width + 3 * img_x + 1] = 255; // G
				image.data[3 * img_y * img_width + 3 * img_x + 2] = 255; // B
			}
  } 
  }
 
}

FFI_PLUGIN_EXPORT void addCross(SickScanPointCloudMsg* pixels) {
    int thickness = 1;
    int midx = image.width / 2;
    int midy = image.height / 2;
    for (int x = midx - 7; x <= midx + 7; x++) {
      // draw horizontal
      for (int y = midy - thickness; y < midy + thickness; y++) {
        image.data[3 * y * image.width + 3 * x + 0] = 0; // B
        image.data[3 * y * image.width + 3 * x + 1] = 0; // G
        image.data[3 * y * image.width + 3 * x + 2] = 255; // R
      }
    }
    for (int y = midy - 7; y <= midy + 7; y++) {
      // draw vertical
      for (int x = midx - thickness; x < midx + thickness; x++) {
        image.data[3 * y * image.width + 3 * x + 0] = 0; // B
        image.data[3 * y * image.width + 3 * x + 1] = 0; // G
        image.data[3 * y * image.width + 3 * x + 2] = 255; // R
      }
    }
  }

  /// Draws a triangle in the area behind lidar that doesn't include data
FFI_PLUGIN_EXPORT void addHiddenArea() {
    /// NEED IMAGE TO BE SQUARE FOR THIS TO WORK
    for (int y = image.height - 1; y > (int)(image.height / 2); y--) {
      for (int x = image.width - y; x < y; x++) {
        image.data[3 * y * image.width + 3 * x + 0] = 130; // R
        image.data[3 * y * image.width + 3 * x + 1] = 130; // G
        image.data[3 * y * image.width + 3 * x + 2] = 130; // B
      }
    }
}


FFI_PLUGIN_EXPORT Image getLatestImage() { 
  return image;
}
